enhancements/installer/openstack-baremetal-compute.md

```diff
@@ -0,0 +1,441 @@
+will function as expected.
+
+We may however, need to add a more complex error recovery and fallback whereby
+we try to recreate the port without these features or possibly even create a
+server without adding an explicit port.
+
+In that case, we should not introduce any new fields to the provider `spec`
+field in the `Machine` object. Instead, the code should either recover from
+errors or do the right thing if the SG/Address Pair fields are empty.
+
+#### Add the necessary CI infrastructure
+
+The [OpenStack cloud we use for CI][moc] does not currently provide Ironic and
+baremetal machines. A considerable portion if this effort will be to ensure
+proper testing.
+
+See the [Test Plan](#test-plan) section for more details.
+
+### Probable issues
+
+* When attaching and detaching, the expected timings must be adjusted to
+  reflect the different boot and shutdown latencies of physical machines
+
+* When attaching, the Certificate Signing Request timeout must be adjusted to
+  reflect the different boot and shutdown latencies of physical machines
+
+### Risks and Mitigations
+
+#### Development and Testing Environments
+
+None of the OpenStack environments we currently use have access to Ironic or
+provide baremetal nodes and we don't have the opportunity to add them easily.
+
+This poses a significant risk to both timely development (dev environment) as
+well as support of the feature (CI).
+
+We propose several options in the [Test Plan](#test-plan) section below.
+
+#### Uncertainty about OpenStack's extra networking resources
+
+The OpenShift Installer as well as [cluster-api-provider-openstack][capo]
+currently create a port with the following properties set for every node:
+
+* Security groups
+* Attached to the network/subnet created during the installation
+* Allowed Address pairs set for the VRRP IP addresses defined by the installer
+* (optional) network trunk
+
+Some of these options (e.g. security groups) are only supported for certain ML2
+drivers. Others (attaching to a given tenant network) only work for the
+`neutron` Ironic networking interface but not for the `flat` one.
+
+It is unclear what happens when the installer tries to set any of these in
+environments whehere they aren't supported. If that results in error, we may
+have to detect these situations, pass them to [CAPO][capo] and modify how we
+create the node and port resources.
+
+This can be mitigated by not using [CAPO][capo] for the baremetal nodes.
+Instead, the OpenShift administrator would create them manually, pass the
+correct Ignition configuration and approve the certificate signing requests
+(CSRs) explicitly.
+
+This manual process of adding compute nodes is described in the [OpenStack
+user-provisioned infrastructure (UPI) document][openstack-upi] and should work
+here as well.
+
+## Design Details
+
+### Test Plan
+
+The OpenStack cloud running CI does not currently support Ironic and bare metal
```

Comment:
Should we document a (preferred) alternative where we have access to a cloud with ironic and BM nodes?

> In the event where we do not have access to a cloud with Ironic service and BM nodes, ...

My concern is that installing OpenStack, then installing OpenShift on top of it, then finally running the test suite might easily get us over the 4h timeout allocated for CI jobs.

Another possibility could be to prepare an OpenStack environment with virtualized BM nodes (as suggested here) but instead of installing it for each job, we'll manually set it up once and give access to it to the CI job, as a cloud platform.

Reply:
++ OpenStack is a fairly stable platform, there should be no real reason you can't just connect to an existing openstack cloud and ask for Bare Metal machines if they are available to perform CI testing. Although, IF you HAVE to change the cloud, then you've likely started to breach an abstraction layer or boundary which raises another question as to why the change needs to take place. What is ultimately driving the change? and is it being made for legitimate reasons would be something any of the upstream OpenStack teams would be able to rightfully able to ask.